src/PixelFlags.cc

```diff
@@ -57,87 +59,153 @@ class FootprintBits : public afw::detection::FootprintFunctor<MaskedImageT> {
     func.apply(footprint);
-    if (func.getBits() & (MaskedImageT::Mask::getPlaneBitMask("EDGE") |
-                          MaskedImageT::Mask::getPlaneBitMask("NO_DATA"))) {
-        _flagHandler.setValue(measRecord, EDGE, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("BAD")) {
-        _flagHandler.setValue(measRecord, BAD, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("INTRP")) {
-        _flagHandler.setValue(measRecord, INTERPOLATED, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("SAT")) {
-        _flagHandler.setValue(measRecord, SATURATED, true);
+
+    // Check to make sure the NO_DATA mask exists
+    try {
+        bitmask = MaskedImageT::Mask::getPlaneBitMask("NO_DATA");
+    } catch (pex::exceptions::InvalidParameterError & err){
+        throw LSST_EXCEPT(FatalAlgorithmError, err.what());
     }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("CR")) {
-        _flagHandler.setValue(measRecord, CR, true);
+
+    // Set the EDGE flag if the bitmask has NO_DATA set
+    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("NO_DATA")){
+        auto iter = _predefinedAnyKeys.find("EDGE");
+        assert(iter != _predefinedAnyKeys.end());
+        measRecord.set(iter->second, true);
     }
 
+    // update the source record for the predefined any keys
+    updateFlags(&_predefinedAnyKeys,&func,measRecord);
+
     // Check for bits set in the 3x3 box around the center
-    afw::geom::Point2I llc(afw::image::positionToIndex(center.getX())-1, afw::image::positionToIndex(center.getY()) - 1);
+    afw::geom::Point2I llc(afw::image::positionToIndex(center.getX())-1,
+                           afw::image::positionToIndex(center.getY()) - 1);
 
     afw::detection::Footprint const middle(afw::geom::BoxI(llc, afw::geom::ExtentI(3))); // central 3x3
     func.apply(middle);
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("INTRP")) {
-        _flagHandler.setValue(measRecord, INTERPOLATED_CENTER, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("SAT")) {
-        _flagHandler.setValue(measRecord, SATURATED_CENTER, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("CR")) {
-        _flagHandler.setValue(measRecord, CR_CENTER, true);
-    }
-    _flagHandler.setValue(measRecord, FAILURE, false);
+
+    // Update the built in flags which have to do with the center of the footprint
+    updateFlags(&_predefinedCenterKeys,&func,measRecord);
+
+    // update all the flags that are defined in the config
+    updateFlags(&_anyKeys, &func, measRecord);
+    updateFlags(&_centerKeys, &func, measRecord);
 }
 
 void PixelFlagsAlgorithm::fail(afw::table::SourceRecord & measRecord, MeasurementError * error) const {
-    _flagHandler.handleFailure(measRecord, error);
+    measRecord.set(_generalFailure,true);
+    assert (error == NULL);
```

Comment:
Because the above is guaranteed never to throw a `MeasurementError`? I think it's worth a comment to explain.